src/JestExt.ts

```diff
@@ -59,119 +61,95 @@ export class JestExt {
           this.channel.appendLine(output)
         }
       })
-      .on('executableStdErr', (error: Buffer) => {
-        const message = error.toString()
-
-        if (this.shouldIgnoreOutput(message)) {
-          return
-        }
-
-        // The "tests are done" message comes through stdErr
-        // We want to use this as a marker that the console should
-        // be cleared, as the next input will be from a new test run.
-        if (this.clearOnNextInput) {
-          this.clearOnNextInput = false
-          this.parsingTestFile = false
-          this.testsHaveStartedRunning()
-        }
-        // thanks Qix, http://stackoverflow.com/questions/25245716/remove-all-ansi-colors-styles-from-strings
-        const noANSI = message.replace(
-          /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g,
-          ''
-        )
-        if (noANSI.includes('snapshot test failed')) {
-          this.detectedSnapshotErrors()
-        }
-
-        this.channel.appendLine(noANSI)
-      })
-      .on('nonTerminalError', (error: string) => {
+      .onJestEditorSupportEvent('executableStdErr', (error: Buffer) => this.handleStdErr(error))
+      .onJestEditorSupportEvent('nonTerminalError', (error: string) => {
         this.channel.appendLine(`Received an error from Jest Runner: ${error.toString()}`)
       })
-      .on('exception', result => {
+      .onJestEditorSupportEvent('exception', result => {
         this.channel.appendLine(`\nException raised: [${result.type}]: ${result.message}\n`)
       })
-      .on('terminalError', (error: string) => {
+      .onJestEditorSupportEvent('terminalError', (error: string) => {
         this.channel.appendLine('\nException raised: ' + error)
       })
+  }
 
+  public startProcess() {
+    if (this.jestProcessManager.numberOfProcesses > 0) {
+      return
+    }
     // The theme stuff
     this.setupDecorators()
     // The bottom bar thing
     this.setupStatusBar()
     //reset the jest diagnostics
     resetDiagnostics(this.failDiagnostics)
 
-    this.forcedClose = false
-    // Go!
-    if (this.pluginSettings.runAllTestsFirst) {
-      this.jestProcess.start(false)
-    } else {
-      this.startWatchMode()
-    }
-  }
+    this.jestProcess = this.jestProcessManager.startJestProcess({
+      watch: true,
+      keepAlive: true,
+      exitCallback: (_, jestProcessInWatchMode) => {
+        if (jestProcessInWatchMode) {
+          this.jestProcess = jestProcessInWatchMode
 
-  public stopProcess() {
-    this.channel.appendLine('Closing Jest jest_runner.')
-    this.closeJest()
-    delete this.jestProcess
-    status.stopped()
-  }
+          this.channel.appendLine('Finished running all tests. Starting watch mode.')
+          status.running('Starting watch mode')
 
-  private closeJest() {
-    if (!this.jestProcess) {
-      return
-    }
-    this.forcedClose = true
-    this.jestProcess.closeProcess()
+          this.assignHandlers(this.jestProcess)

```

Comment:
When watch mode fails, this output is only assigned once.

Reply:
You might be right. Let me investigate this.

Reply:
You are right @seanpoulter. So I want to avoid adding more callbacks and taken that we already have separate `JestProcess` to deal with this kind of things, I will record registered callbacks and restore them when process is restarted inside `JestProcess` class. Feels like a right place to handle that. I will have it tomorrow.